osquery/tables/system/windows/shellbags.cpp

```diff
@@ -0,0 +1,441 @@
+          (shell_data.find("BBAF933B") != std::string::npos) ||
+          (shell_data.find("00EEBEBE") != std::string::npos)) {
+        build_shellbag.push_back("[USER PROPERTY VIEW]\\");
+        std::string full_path = buildPath(build_shellbag);
+        full_path.pop_back();
+        r["path"] = full_path;
+        results.push_back(r);
+        return;
+      }
+      std::vector<size_t> wps_list;
+      // check if Windows Property Store version 1SPS is in shellbag data, there
+      // could be multiple
+      size_t wps = shell_data.find("31535053");
+      while (wps != std::string::npos) {
+        wps_list.push_back(wps);
+        wps = shell_data.find("31535053", wps + 1);
+      }
+      std::string property_name = propertyStore(shell_data, wps_list);
+      build_shellbag.push_back(property_name + "\\");
+      std::string full_path = buildPath(build_shellbag);
+      full_path.pop_back();
+      r["path"] = full_path;
+      results.push_back(r);
+      return;
+    } else {
+      LOG(WARNING) << "Unsupported Shellbag format: " << shell_data;
+      build_shellbag.push_back("[UNSUPPORTED FORMAT]\\");
+      std::string full_path = buildPath(build_shellbag);
+      full_path.pop_back();
+      r["path"] = full_path;
+      results.push_back(r);
+      return;
+    }
+  }
+
+  if (file_entry.path == "[UNSUPPORTED SHELL EXTENSION]\\") {
+    build_shellbag.push_back(file_entry.path);
+    std::string full_path = buildPath(build_shellbag);
+    full_path.pop_back();
+    r["path"] = full_path;
+    results.push_back(r);
+    return;
+  }
+
+  r["modified_time"] = INTEGER(file_entry.dos_modified);
+  r["created_time"] = INTEGER(file_entry.dos_created);
+  r["accessed_time"] = INTEGER(file_entry.dos_accessed);
+
+  build_shellbag.push_back(file_entry.path + "\\");
+  long long mft_entry = file_entry.mft_entry;
+  int mft_sequence = file_entry.mft_sequence;
+  std::string full_path = buildPath(build_shellbag);
+  full_path.pop_back();
+
+  r["path"] = full_path;
+  r["mft_entry"] = BIGINT(mft_entry);
+  r["mft_sequence"] = INTEGER(mft_sequence);
+  results.push_back(r);
+}
+
+// Recursively loop through shellbag entries in the Registry
+void parseShellbags(const std::string& path,
+                    std::vector<std::string>& build_shellbag,
+                    QueryData& results,
+                    const std::string& sid,
+                    const std::string& source) {
+  QueryData shellbag_data;
+  queryKey(path, shellbag_data);
+  for (const auto& rKey : shellbag_data) {
+    auto key_type = rKey.find("type");
+    auto key_path = rKey.find("path");
+    if (key_type == rKey.end() || key_path == rKey.end()) {
+      continue;
+    }
+    // For Shellbags Reg keys the last character is a number
+    if (!isdigit(key_path->second.back())) {
+      continue;
+    }
+    if (rKey.at("data") == "") {
+      continue;
+    }
+    parseShellData(rKey.at("data"), build_shellbag, results, sid, source);
```

Comment:
Consider `[]` or `find`.

Reply:
could you clarify a bit by what you mean by `[]` or `find`?  
Do u mean something like putting the string values "rKey.at(data), sid, source" in an array and passing three values instead of five?